### PR TITLE
[mlir][python] add binding to `#gpu.object`

### PR DESCRIPTION
**Pull request by makslevental**
The goal here is to enable extracting the compiled artifact for downstream (GPU) compilers ([currently you have to do some unholy things...](https://github.com/makslevental/mlir-python-extras/blob/f12e5d1cbc69e2d5c689f05e33a2fbc03890bd17/examples/cuda_matmul_opt.py#L35-L37)).

### CODE REVIEW

File: mlir/lib/Bindings/Python/DialectGPU.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+          "get",
+          [](py::object cls, MlirAttribute target, uint32_t format,
+             py::bytes object, std::optional<MlirAttribute> mlirObjectProps,
+             std::optional<MlirContext> context) {
+            py::buffer_info info(py::buffer(object).request());
+            MlirStringRef objectStrRef =
+                mlirStringRefCreate(static_cast<char *>(info.ptr), info.size);
+            return cls(mlirGPUObjectAttrGet(
+                context.has_value() ? *context
+                                    : mlirAttributeGetContext(target),
```
Comment (ftynse) [inline mlir/lib/Bindings/Python/DialectGPU.cpp:44]: Is there a case where we want a different context than the one `target` lives in?

File: mlir/lib/CAPI/Dialect/GPU.cpp [hunk 2]
```diff
@@ -9,5 +9,57 @@
 #include "mlir-c/Dialect/GPU.h"
 #include "mlir/CAPI/Registration.h"
 #include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "llvm/Support/Casting.h"
 
 MLIR_DEFINE_CAPI_DIALECT_REGISTRATION(GPU, gpu, mlir::gpu::GPUDialect)
+
```
Comment (ftynse) [inline mlir/lib/CAPI/Dialect/GPU.cpp:15]: Nit: can we do `using namespace mlir` and drop the obnoxious prefix below?

**Review by ftynse — APPROVED** [review]

**Review by jpienaar — APPROVED** [review]

**Review by grypp — APPROVED** [review]
Not an expert in py bindings fields. But PR makes setting target nicer and simpler

File: mlir/lib/Bindings/Python/DialectGPU.cpp (hunk 1 above)
Comment (makslevental) [reply mlir/lib/Bindings/Python/DialectGPU.cpp:44]: Ya I guess not

**Pull request merged**